### PR TITLE
Hooks documentation

### LINKED ISSUE #1456: Document webhooks API and usage

### PR DESCRIPTION
**Pull request by itaiad200**
closes #1456

### CODE REVIEW

File: docs/reference/hooks.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
+---
+layout: default
+title: Hooks
+parent: Reference
```
Comment (ozkatz) [inline docs/reference/hooks.md:4]: I would suggest having this as a top-level item in the docs tree

File: docs/reference/hooks.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+ 
+lakeFS `Actions` are handled per repository and cannot be reused.  
+Failure of any `Hook` under any `Action` will result in aborting the lakeFS operation that is taking place.
+
+## Action
+An `Action` is a list of `Hooks` with the same trigger configuration.
+`Hooks` list under an `Action` is ordered and an `Hook` will only be executed if all previous `Hooks` passed.  
+
+### Hook
+An `Hook` is the basic building block of an `Action`.
```
Comment (ozkatz) [inline docs/reference/hooks.md:24]:
```suggestion
A `Hook` is the basic building block of an `Action`. 
```

File: docs/reference/hooks.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+lakeFS `Actions` are handled per repository and cannot be reused.  
+Failure of any `Hook` under any `Action` will result in aborting the lakeFS operation that is taking place.
+
+## Action
+An `Action` is a list of `Hooks` with the same trigger configuration.
+`Hooks` list under an `Action` is ordered and an `Hook` will only be executed if all previous `Hooks` passed.  
+
+### Hook
+An `Hook` is the basic building block of an `Action`. 
+Failure of a single `Hook` will stop the execution of the containing `Action` and fail the `Run`.
```
Comment (ozkatz) [inline docs/reference/hooks.md:25]: This is true for the two types of hooks currently supported. Obviously i won't be the case for `post-merge`, will it? I suggest documenting this above when explaining the current types.

File: docs/reference/hooks.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+## Action
+An `Action` is a list of `Hooks` with the same trigger configuration.
+`Hooks` list under an `Action` is ordered and an `Hook` will only be executed if all previous `Hooks` passed.  
+
+### Hook
+An `Hook` is the basic building block of an `Action`. 
+Failure of a single `Hook` will stop the execution of the containing `Action` and fail the `Run`. 
+
+### Action file
+Schema of the Action file:
```
Comment (ozkatz) [inline docs/reference/hooks.md:28]: At this point in reading the doc, I don't know what an Action file is - perhaps explain the concept in a seperate paragraph prior to this one (hooks are managed by writing to prefix in the lakeFS repository, this allows configuration-as-code inside lakeFS, action files are declerative and written in YAML, ...). After such a paragraph, this one would make sense :)

File: docs/reference/hooks.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+|Property          |Description                                            |Required |Default Value
+|------------------|-------------------------------------------------------|---------|--------------------------------------|
+|name              |Identify the Action file                               |false    | If missing, filename is used instead 
+|on                |List of events that will trigger the hooks             |true     |
```
Comment (ozkatz) [inline docs/reference/hooks.md:33]: please specify data type (string? array of strings? enum of certain values?)

File: docs/reference/hooks.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+      url: "https://your.domain.io/webhook?notmp=true?t=1za2PbkZK1bd4prMuTDr6BeEQwWYcX2R"
+  - id: no_freeze
+    type: webhook
+    description: check production is not in dev freeze
+    properties:
+      url: "https://your.domain.io/webhook?nofreeze=true?t=1za2PbkZK1bd4prMuTDr6BeEQwWYcX2R"
+```
+
+## Run
+`Run` is the instantiation of the repo `Action` files when the event occurs.
```
Comment (ozkatz) [inline docs/reference/hooks.md:63]:
```suggestion
A `Run` is an instantiation of the repository's `Action` files when the triggering event occurs. For example, If our repository contains a pre-commit hook, every commit would generate a `Run` for that specific commit. 
```

File: docs/reference/hooks.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+## How to upload Action files
+`Action` files should be uploaded with the prefix `_lakefs_actions/` to the lakeFS repository.
+When an actionable event (see Supported Events above) takes place, lakeFS will read all files with prefix `_lakefs_actions/`
+in the repository branch where the action occurred. 
+A failure to parse an `Action` file will result with the event aborted. 
+ 
+For example, lakeFS will search and execute all matching `Action` files with the prefix `lakefs://repo1/feature-1/_lakefs_actions/` on:
+1. Commit to `feature-1` branch on `repo1` repository. 
+2. Merge to `master` branch from `feature-1` branch on `repo1` repository.
```
Comment (ozkatz) [inline docs/reference/hooks.md:75]:
```suggestion
2. Merge to `main` branch from `feature-1` branch on `repo1` repository.
```

File: docs/reference/hooks.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+Example:
+```yaml
+name: Good files check
+description: set of checks to verify that branch is good
+on:
+  pre-commit:
+  pre-merge:
+    branches:
+      - master
```
Comment (ozkatz) [inline docs/reference/hooks.md:48]:
```suggestion
      - main
```

File: docs/reference/hooks.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+1. Commit to `feature-1` branch on `repo1` repository. 
+2. Merge to `master` branch from `feature-1` branch on `repo1` repository.
+
+## Runs API
+OpenAPI endpoint exposes the results of `Runs` execution per repository, branch, commit and specific `Action`.
+The endpoint also allows to download the execution log of any executed `Hook` under each `Run` for observability.
+
+### Run results objectstore files 
+There are 2 types of files that are stored in the metadata section of lakeFS repository with each `Run`:
+1. `_lakefs/actions/log/<<runID>>/<<hookRunID>>.log` - Execution log of the specific `Hook` run.
```
Comment (ozkatz) [inline docs/reference/hooks.md:83]: Why the double brackets? is that a markdown thing?

File: docs/reference/hooks.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+2. `_lakefs/actions/log/<<runID>>/run.manifest` - Manifest with all `Hooks` execution for the run with their results and additional metadata.
+
+## Type of hooks
+Currently, there's only a single type of `Hooks` that is supported by lakeFS, the `webhook`.
```
Comment (ozkatz) [inline docs/reference/hooks.md:87]:
```suggestion
Currently, there's only a single type of `Hooks` that is supported by lakeFS: `webhook`.
```

File: docs/reference/hooks.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+## Webhooks
+A `Webhook` is an `Hook` type that sends an HTTP POST request to the configured URL.
```
Comment (ozkatz) [inline docs/reference/hooks.md:90]:
```suggestion
A `Webhook` is a `Hook` type that sends an HTTP POST request to the configured URL.
```

File: docs/reference/hooks.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+Any non 2XX response by the responding endpoint will fail the `Hook`, cancel the execution of the following `Hooks` 
+under the same `Action` and abort the operation that triggered the `Action`.
+
+**Warning:** Actions Run is a blocking operation - users should not use the webhook for long-running tasks (e.g. Running Spark jobs and waiting to completion).
+Moreover, since the branch is locked during the execution, any write operation to the branch (like uploading file or committing) by the webhook server is bound to create a deadlock.
```
Comment (ozkatz) [inline docs/reference/hooks.md:95]: AFAIU, it currently won't deadlock, but will instead simply fail

File: docs/reference/hooks.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+{: .note } 
+
+### Action file Webhook properties
+
+|Property          |Description                                            |Required |Default Value
+|------------------|-------------------------------------------------------|---------|--------------------------------------|
```
Comment (ozkatz) [inline docs/reference/hooks.md:101]: Please add data type

File: docs/reference/hooks.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+  "repository_id": "repo1",
+  "branch_id": "feature-1",
+  "source_ref": "feature-1",
+  "commit_message": "merge commit message",
+  "committer": "committer",
+  "commit_metadata": {
+    "key": "value"
+  }
+}
+```
```
Comment (ozkatz) [inline docs/reference/hooks.md:154]: Missing a link to the API reference and the CLI hooks sub-command. 
Also, I'd add another paragraph at the beginning explaining some of the use cases for this (you can take example from the lakeFS-hooks repo)

**Review by ozkatz — CHANGES_REQUESTED** [review]
To change: data types for different YAML values, and missing a high level overview before diving into the schema of Action files..

File: docs/reference/hooks.md (hunk 1 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+When an actionable event (see Supported Events above) takes place, lakeFS will read all files with prefix `_lakefs_actions/`
+in the repository branch where the action occurred. 
+A failure to parse an `Action` file will result with the event aborted. 
+ 
+For example, lakeFS will search and execute all matching `Action` files with the prefix `lakefs://repo1/feature-1/_lakefs_actions/` on:
+1. Commit to `feature-1` branch on `repo1` repository. 
+2. Merge to `master` branch from `feature-1` branch on `repo1` repository.
+
+## Runs API
+OpenAPI endpoint exposes the results of `Runs` execution per repository, branch, commit and specific `Action`.
```
Comment (johnnyaug) [inline docs/reference/hooks.md:78]: This section is not clear. Need to rephrase and elaborate.

File: docs/reference/hooks.md [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+The endpoint also allows to download the execution log of any executed `Hook` under each `Run` for observability.
+
+### Run results objectstore files 
+There are 2 types of files that are stored in the metadata section of lakeFS repository with each `Run`:
+1. `_lakefs/actions/log/<<runID>>/<<hookRunID>>.log` - Execution log of the specific `Hook` run.
+2. `_lakefs/actions/log/<<runID>>/run.manifest` - Manifest with all `Hooks` execution for the run with their results and additional metadata.
+
+## Type of hooks
```
Comment (johnnyaug) [inline docs/reference/hooks.md:86]:
```suggestion
## Hook Types
```

File: docs/reference/hooks.md (hunk 3 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Right, Done.

File: docs/reference/hooks.md (hunk 8 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+      url: "http://<host:port>/webhooks/schema"
+      timeout: 1m30s
+      query_params:
+        disallow: ["user_", "private_"]
+        prefix: public/
+...
+```
+
+### Request body schema
+A JSON object with the following fields:
```
Comment (johnnyaug) [inline docs/reference/hooks.md:123]: Upon
```suggestion
Upon execution, a webhook will send a request containing a JSON object with the following fields:
```

File: docs/reference/hooks.md (hunk 7 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md (hunk 10 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md (hunk 11 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md (hunk 12 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+Supported Events:
+1. `pre_commit` - Action runs when the commit occurs, before the commit is finalized.
+2. `pre_merge` - Action runs when the merge occurs, before the merge is finalized.
+ 
+lakeFS `Actions` are handled per repository and cannot be reused.  
+Failure of any `Hook` under any `Action` will result in aborting the lakeFS operation that is taking place.
+
+## Action
+An `Action` is a list of `Hooks` with the same trigger configuration.
+`Hooks` list under an `Action` is ordered and an `Hook` will only be executed if all previous `Hooks` passed.
```
Comment (johnnyaug) [inline docs/reference/hooks.md:21]:
```suggestion
The `Hooks` under an `Action` are ordered, and a `Hook` will only be executed if all previous `Hooks` had passed.
```

File: docs/reference/hooks.md [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+    type: webhook
+    description: check production is not in dev freeze
+    properties:
+      url: "https://your.domain.io/webhook?nofreeze=true?t=1za2PbkZK1bd4prMuTDr6BeEQwWYcX2R"
+```
+
+## Run
+`Run` is the instantiation of the repo `Action` files when the event occurs.
+lakeFS will fetch, parse and filter the repository `Action` files and start to execute the `Hooks` under each `Action`.
+All executed `Hooks` (each with `hook_run_id`) exists in the context of that `Run` (`run_id`).
```
Comment (johnnyaug) [inline docs/reference/hooks.md:65]: This is too low-level IMHO.
```suggestion
Once an event occurs, the `Hooks` from the relevant `Action` files are executed.
```

File: docs/reference/hooks.md (hunk 4 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+For example, lakeFS will search and execute all matching `Action` files with the prefix `lakefs://repo1/feature-1/_lakefs_actions/` on:
+1. Commit to `feature-1` branch on `repo1` repository. 
+2. Merge to `master` branch from `feature-1` branch on `repo1` repository.
+
+## Runs API
+OpenAPI endpoint exposes the results of `Runs` execution per repository, branch, commit and specific `Action`.
+The endpoint also allows to download the execution log of any executed `Hook` under each `Run` for observability.
+
+### Run results objectstore files 
+There are 2 types of files that are stored in the metadata section of lakeFS repository with each `Run`:
```
Comment (johnnyaug) [inline docs/reference/hooks.md:82]: Maybe clarify that the metadata section lives in the underlying storage and not the repository.

File: docs/reference/hooks.md (hunk 5 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md (hunk 6 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md (hunk 9 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Removed.

File: docs/reference/hooks.md [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+## Run
+`Run` is the instantiation of the repo `Action` files when the event occurs.
+lakeFS will fetch, parse and filter the repository `Action` files and start to execute the `Hooks` under each `Action`.
+All executed `Hooks` (each with `hook_run_id`) exists in the context of that `Run` (`run_id`).   
+
+## How to upload Action files
+`Action` files should be uploaded with the prefix `_lakefs_actions/` to the lakeFS repository.
+When an actionable event (see Supported Events above) takes place, lakeFS will read all files with prefix `_lakefs_actions/`
+in the repository branch where the action occurred. 
+A failure to parse an `Action` file will result with the event aborted.
```
Comment (johnnyaug) [inline docs/reference/hooks.md:71]: Again, not applicable to hooks in general

**Review by johnnyaug — CHANGES_REQUESTED** [review]
Very good documentation.

File: docs/reference/hooks.md (hunk 13 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md (hunk 14 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+---
+
+# Configurable Hooks
+Like other version control systems, lakeFS allows the configuration of `Actions` to trigger when predefined events occur.
+ 
+Supported Events:
+1. `pre_commit` - Action runs when the commit occurs, before the commit is finalized.
+2. `pre_merge` - Action runs when the merge occurs, before the merge is finalized.
+ 
+lakeFS `Actions` are handled per repository and cannot be reused.
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply docs/reference/hooks.md]: Yes, rephrased.

File: docs/reference/hooks.md [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+
+# Configurable Hooks
+Like other version control systems, lakeFS allows the configuration of `Actions` to trigger when predefined events occur.
+ 
+Supported Events:
+1. `pre_commit` - Action runs when the commit occurs, before the commit is finalized.
+2. `pre_merge` - Action runs when the merge occurs, before the merge is finalized.
+ 
+lakeFS `Actions` are handled per repository and cannot be reused.  
+Failure of any `Hook` under any `Action` will result in aborting the lakeFS operation that is taking place.
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+ 
+For example, lakeFS will search and execute all matching `Action` files with the prefix `lakefs://repo1/feature-1/_lakefs_actions/` on:
+1. Commit to `feature-1` branch on `repo1` repository. 
+2. Merge to `master` branch from `feature-1` branch on `repo1` repository.
+
+## Runs API
+OpenAPI endpoint exposes the results of `Runs` execution per repository, branch, commit and specific `Action`.
+The endpoint also allows to download the execution log of any executed `Hook` under each `Run` for observability.
+
+### Run results objectstore files
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md (hunk 19 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Need the different run ids explanations for the API to make sense.

File: docs/reference/hooks.md (hunk 20 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/reference/hooks.md (hunk 21 above)
Comment (itaiad200) [reply docs/reference/hooks.md]: Done.

File: docs/hooks.md [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+1. Format Validator:
+   A webhook that checks new files to ensure they are of a set of allowed data format.
+2. Schema Validator:
+   A webhook that reads new Parquet and ORC files to ensure they don't contain a block list of column names (or name prefixes). 
+   This is useful when we want to avoid accidental PII exposure.
+
+For more examples and configuration samples, check out [lakeFS-hooks](https://github.com/treeverse/lakeFS-hooks) example repo.
+
+## Action
+An `Action` is a list of `Hooks` with the same trigger configuration.
```
Comment (nopcoder) [inline docs/hooks.md:31]: maybe mention 'event' - which triggers the action

**Review by nopcoder — APPROVED** [review]
Minor comment - great docs!

File: docs/hooks.md [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+2. Schema Validator:
+   A webhook that reads new Parquet and ORC files to ensure they don't contain a block list of column names (or name prefixes). 
+   This is useful when we want to avoid accidental PII exposure.
+
+For more examples and configuration samples, check out [lakeFS-hooks](https://github.com/treeverse/lakeFS-hooks) example repo.
+
+## Action
+An `Action` is a list of `Hooks` with the same trigger configuration, i.e. an event will trigger all `Hooks` under an `Action`, or none at all.
+The `Hooks` under an `Action` are ordered, and a `Hook` will only be executed if all previous `Hooks` had passed.
+
```
Comment (johnnyaug) [inline docs/hooks.md:33]: These two sentences are contradictory

File: docs/hooks.md [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+  - id: no_freeze
+    type: webhook
+    description: check production is not in dev freeze
+    properties:
+      url: "https://your.domain.io/webhook?nofreeze=true?t=1za2PbkZK1bd4prMuTDr6BeEQwWYcX2R"
+```
+
+## Run
+A `Run` is an instantiation of the repository's `Action` files when the triggering event occurs. 
+For example, If our repository contains a pre-commit hook, every commit would generate a `Run` for that specific commit.
```
Comment (johnnyaug) [inline docs/hooks.md:75]:
```suggestion
For example, if our repository contains a pre-commit hook, every commit would generate a `Run` for that specific commit.
```

**Review by johnnyaug — APPROVED** [review]

File: docs/hooks.md (hunk 26 above)
Comment (itaiad200) [reply docs/hooks.md:33]: I understand why it's confusing: trigger != execution .
Added clarification that execution depends on failures, not triggering.

**Review by ozkatz — APPROVED** [review]

File: docs/hooks.md [hunk 28]
```diff
@@ -0,0 +1,172 @@
...
+  pre-commit:
+  pre-merge:
+    branches:
+      - main
+hooks:
+  - id: no_temp
+    type: webhook
+    description: checking no temporary files found
+    properties:
+      url: "https://your.domain.io/webhook?notmp=true?t=1za2PbkZK1bd4prMuTDr6BeEQwWYcX2R"
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply docs/hooks.md:65]: There's an example of it below. I want this example to focus on the structure of the Action file with the hooks, less about configuring the specific webhook implementation.

**Pull request merged**